### PR TITLE
[Python] Keep reference to callback function on timeout

### LINKED ISSUE #30870: [BUG] [Python] Calling GetConnectedDeviceSync with timeout leads to segfault
### Reproduction steps

Calling `GetConnectedDeviceSync` with `timeoutMs=30000` leads to a crash:

```
Thread 4 "matter-server" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fb37fe1e0 (LWP 1205)]
0x0000007fbe8ed494 in _PyEval_EvalFrameDefault () from /usr/local/bin/../lib/libpython3.11.so.1.0
(gdb) bt
#0  0x0000007fbe8ed494 in _PyEval_EvalFrameDefault () from /usr/local/bin/../lib/libpython3.11.so.1.0
#1  0x0000007fbe8eabb0 in ?? () from /usr/local/bin/../lib/libpython3.11.so.1.0
#2  0x0000007fbe8fd98c in PyObject_Vectorcall () from /usr/local/bin/../lib/libpython3.11.so.1.0
#3  0x0000007fbcd45d9c in ?? () from /usr/local/lib/python3.11/lib-dynload/_ctypes.cpython-311-aarch64-linux-gnu.so
#4  0x0000007fbcd20e14 in ?? () from /usr/lib/aarch64-linux-gnu/libffi.so.7
#5  0x0000007fbcd211c4 in ?? () from /usr/lib/aarch64-linux-gnu/libffi.so.7
#6  0x0000007fb98b8f80 in (anonymous namespace)::GetDeviceCallbacks::OnConnectionFailureFn (context=0x7fa40274b0, peerId=..., error=...) at ../src/controller/python/ChipDeviceController-ScriptBinding.cpp:742
#7  0x0000007fb9c5adcc in chip::OperationalSessionSetup::NotifyConnectionCallbacks (failureReady=..., successReady=..., error=..., peerId=..., performingAddressUpdate=false, exchangeMgr=0x55a6fa7430, optionalSessionHandle=...) at ../src/app/OperationalSessionSetup.cpp:338
#8  0x0000007fb9c5ac50 in chip::OperationalSessionSetup::DequeueConnectionCallbacks (this=0x7fa402b580, error=..., releaseBehavior=chip::OperationalSessionSetup::ReleaseBehavior::Release) at ../src/app/OperationalSessionSetup.cpp:314
#9  0x0000007fb9c5bbbc in chip::OperationalSessionSetup::OnNodeAddressResolutionFailed (this=0x7fa402b580, peerId=..., reason=...) at ../src/app/OperationalSessionSetup.cpp:582
#10 0x0000007fb9e74fc4 in chip::AddressResolve::Impl::Resolver::HandleAction (this=0x7fba03a5f0 <chip::AddressResolve::Resolver::Instance()::gResolver>, current=...) at ../src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:333
#11 0x0000007fb9e750dc in chip::AddressResolve::Impl::Resolver::HandleTimer (this=0x7fba03a5f0 <chip::AddressResolve::Resolver::Instance()::gResolver>) at ../src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:353
#12 0x0000007fb9e73e64 in chip::AddressResolve::Impl::Resolver::OnResolveTimer (layer=0x7fba036d08 <chip::DeviceLayer::SystemLayerImpl()::gSystemLayerImpl>, context=0x7fba03a5f0 <chip::AddressResolve::Resolver::Instance()::gResolver>) at ../src/lib/address_resolve/AddressResolve_DefaultImpl.h:185
#13 0x0000007fb9cc5420 in chip::System::TimerData::Callback::Invoke (this=0x7fb37fd8c0) at ../src/system/SystemTimer.h:61
#14 0x0000007fb9cc708c in chip::System::TimerPool<chip::System::TimerList::Node>::Invoke (this=0x7fba037310 <chip::DeviceLayer::SystemLayerImpl()::gSystemLayerImpl+1544>, timer=0x55a755a950) at ../src/system/SystemTimer.h:237
#15 0x0000007fb9cc6bb0 in chip::System::LayerImplSelect::HandleEvents (this=0x7fba036d08 <chip::DeviceLayer::SystemLayerImpl()::gSystemLayerImpl>) at ../src/system/SystemLayerImplSelect.cpp:678
#16 0x0000007fb9dfdeb4 in chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::_RunEventLoop (this=0x7fba039ba8 <chip::DeviceLayer::PlatformManagerImpl::sInstance+16>) at ../src/include/platform/internal/GenericPlatformManagerImpl_POSIX.ipp:222
#17 0x0000007fb9dfc958 in chip::DeviceLayer::PlatformManager::RunEventLoop (this=0x7fba039b98 <chip::DeviceLayer::PlatformManagerImpl::sInstance>) at ../src/include/platform/PlatformManager.h:403
#18 0x0000007fb9dfe474 in chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::EventLoopTaskMain (arg=0x7fba039ba8 <chip::DeviceLayer::PlatformManagerImpl::sInstance+16>) at ../src/include/platform/internal/GenericPlatformManagerImpl_POSIX.ipp:253
#19 0x0000007fbe59b648 in start_thread () from /lib/aarch64-linux-gnu/libpthread.so.0
#20 0x0000007fbe696fdc in ?? () from /lib/aarch64-linux-gnu/libc.so.6
quit)
```

### Bug prevalence

Whenever I do this

### GitHub hash of the SDK that was being used

v1.2.0.1 (181b0cb14ff007ec912f2ba6627e05dfb066c008)

### Platform

python

### Platform Version(s)

_No response_

### Anything else?

It seems that the callback function `DeviceAvailableCallback` passed in `GetConnectedDeviceSync` is being GC'ed. Adding a `ctypes.pythonapi.Py_IncRef(ctypes.py_object(DeviceAvailableCallback))` helps to alleviate the problem, but decrementing is not really possible from inside the function :cry:

### PR DESCRIPTION
**Pull request by agners**
When using a timeout when calling GetConnectedDeviceSync() the callback function DeviceAvailableCallback() gets potentially GC'ed. Make sure we hold a reference to the instance.

Fixes: #30870

### CODE REVIEW

**Review by cecille — COMMENTED** [review]
Does this also happen to the other callbacks?

**Review by yunhanw-google — APPROVED** [review]

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 1]
```diff
@@ -759,25 +759,30 @@ def GetConnectedDeviceSync(self, nodeid, allowPASE=True, timeoutMs: int = None):
...
+                    returnDevice = c_void_p(device)
+                    returnErr = err
+                    deviceAvailableCV.notify_all()
+                ctypes.pythonapi.Py_DecRef(ctypes.py_object(self))
+
+        closure = DeviceAvailableClosure()
+        ctypes.pythonapi.Py_IncRef(ctypes.py_object(closure))
         self._ChipStack.Call(lambda: self._dmLib.pychip_GetConnectedDeviceByNodeId(
-            self.devCtrl, nodeid, DeviceAvailableCallback), timeoutMs).raise_on_error()
+            self.devCtrl, nodeid, ctypes.py_object(closure), closure.DeviceAvailableCallback),
```
[the remark this answers is not in the document]
Comment (agners) [reply src/controller/python/chip/ChipDeviceCtrl.py:784]: Uh, that is not a dumb question at all. I misssed calling the callback with the context :see_no_evil: I wonder how that worked :thinking:  I guess the reference counting worked nonetheless so my test case was solved.

I've added another commit which actually uses the context in the callbacks. The context here is the object itself, which is the first argument of the callback functions (self).

**Comment by agners** [comment]
> Does this also happen to the other callbacks?

I did search through the code to find other implementation, but I did not find another problematic cases.

Most of the callbacks are implemented in slightly different ways: E.g. the commission callbacks are part of long-lived object of the type `ChipDeviceController` (see `src/controller/python/chip/ChipDeviceCtrl.py`).

In `src/controller/python/chip/clusters/Attribute.py` the ref counter of transactions objects gets incremented manually in `WriteAttributes()` and `Read()`.

A quite similar situation I found in `src/controller/python/chip/ble/scan_devices.py`. My fix took  inspiration from that case. I did not made global Callback function since from what I understand that is not really necessary.

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tehampson) [reply src/controller/python/chip/ChipDeviceCtrl.py:784]: Yeah, I was very confused as to how this was working, and TBH I am still confused as to how this is working. Even in the reference [implementation you are stating](https://github.com/project-chip/connectedhomeip/pull/30877#issuecomment-1850911899) in `src/controller/python/chip/ble/scan_devices.py` they are not providing the callback as `closure.ScanFoundCallback` but are providing a python function itself like `ScanFoundCallback`, where that function itself calls `closure.DeviceFound(..)`

To keep it consistent among callback that I see in python, can you do something similar? Or can you explain why providing `closure.DeviceAvailableCallback` worked with how you have code now and how it was written before when you were not giving mContext?

Please feel free to mention me in your response as that notifies me much faster

[the remark this answers is not in the document]
Comment (agners) [reply src/controller/python/chip/ChipDeviceCtrl.py:784]: @tehampson  in Python, as it is quite common in object oriented languages, a method is essentially a function with the object reference as it's first argument. In Python this is even explicitly stated with the `self` argument. It seems that the `CFUNCTYPE` annotation allow to directly annotate methods.

That said, I agree that this is not the consistent style. Using a separate function separates the ctypes interface and the Python world nicer as well. I've adjusted the source accordingly.

**Review by tehampson — APPROVED** [review]

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tehampson) [reply src/controller/python/chip/ChipDeviceCtrl.py]: I get that, I just don't understand how it worked for you originally when you didn't provide that argument. That is why I somewhat prefer what you have updated to now. It's that unknown of why it worked when you didn't provide that first argument that took me for a little bit of a ride.

Anyways LGTM now. Thanks

**Pull request merged**